File: bench/bench_ipc.cpp
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <vector>
#include <thread>
#include <atomic>
#include <chrono>
#include <random>

#include "libipc/ipc.h"

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

static std::atomic<bool> g_ready{false};
static std::atomic<bool> g_done{false};

static void wait_ready() {
    while (!g_ready.load(std::memory_order_acquire))
        std::this_thread::yield();
}

struct Stats {
    double total_ms;
    std::size_t count;
    double us_per_datum() const { return (total_ms * 1000.0) / count; }
};

// ---------------------------------------------------------------------------
// ipc::route  —  1 sender, N receivers  (random msg_lo–msg_hi bytes × count)
// ---------------------------------------------------------------------------

static Stats bench_route(int n_receivers, std::size_t count,
                         std::size_t msg_lo, std::size_t msg_hi) {
    const char* name = "bench_route";

    std::vector<std::thread> threads;

    // prepare random payloads
    std::mt19937 rng(42);
    std::uniform_int_distribution<std::size_t> dist(msg_lo, msg_hi);
    std::vector<std::size_t> sizes(count);
    for (auto& s : sizes) s = dist(rng);
    std::vector<char> payload(msg_hi, 'X');

    // sender (created first so shm exists for receivers)
    ipc::route sender(name, ipc::sender);

    // receivers
    for (int i = 0; i < n_receivers; ++i) {
        threads.emplace_back([&, name] {
            ipc::route r(name, ipc::receiver);
            wait_ready();
            while (!g_done.load(std::memory_order_acquire)) {
                auto buf = r.recv(100);
            }
        });
    }

    // let receivers connect
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    g_ready.store(true, std::memory_order_release);

    auto t0 = std::chrono::steady_clock::now();

    for (std::size_t i = 0; i < count; ++i)
        sender.send(payload.data(), sizes[i]);

    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    // signal done, disconnect sender to unblock receivers, then join
    g_done.store(true, std::memory_order_release);
    sender.disconnect();
    for (auto& t : threads) t.join();

    g_ready.store(false);
    g_done.store(false);

    return {ms, count};
}

// ---------------------------------------------------------------------------
// ipc::channel  —  pattern  (random msg_lo–msg_hi bytes × count)
//   pattern: "1-N"  = 1 sender,  N receivers
//            "N-1"  = N senders, 1 receiver
//            "N-N"  = N senders, N receivers
// ---------------------------------------------------------------------------

static Stats bench_channel(const std::string& pattern, int n,
                           std::size_t count, std::size_t msg_lo,
                           std::size_t msg_hi) {
    const char* name = "bench_chan";

    int n_senders   = (pattern == "N-1" || pattern == "N-N") ? n : 1;
    int n_receivers = (pattern == "1-N" || pattern == "N-N") ? n : 1;

    std::size_t per_sender = count / static_cast<std::size_t>(n_senders);

    // prepare random payloads
    std::mt19937 rng(42);
    std::uniform_int_distribution<std::size_t> dist(msg_lo, msg_hi);
    std::vector<std::size_t> sizes(count);
    for (auto& s : sizes) s = dist(rng);
    std::vector<char> payload(msg_hi, 'X');

    // a "control" channel to keep shm alive; also used to disconnect receivers
    ipc::channel ctrl(name, ipc::sender);

    std::vector<std::thread> recv_threads;

    // receivers
    for (int i = 0; i < n_receivers; ++i) {
        recv_threads.emplace_back([&, name] {
            ipc::channel ch(name, ipc::receiver);
            wait_ready();
            while (!g_done.load(std::memory_order_acquire)) {
                auto buf = ch.recv(100);
            }
        });
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    g_ready.store(true, std::memory_order_release);

    auto t0 = std::chrono::steady_clock::now();

    // senders
    std::vector<std::thread> sender_threads;
    for (int s = 0; s < n_senders; ++s) {
        sender_threads.emplace_back([&, s, per_sender, name] {
            ipc::channel ch(name, ipc::sender);
            std::size_t base = static_cast<std::size_t>(s) * per_sender;
            for (std::size_t i = 0; i < per_sender; ++i)
                ch.send(payload.data(), sizes[base + i]);
        });
    }
    for (auto& t : sender_threads) t.join();

    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    g_done.store(true, std::memory_order_release);
    ctrl.disconnect();
    for (auto& t : recv_threads) t.join();

    g_ready.store(false);
    g_done.store(false);

    return {ms, count};
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------

static void print_header(const char* title) {
    printf("\n=== %s ===\n", title);
}

int main(int argc, char** argv) {
    int max_threads = 8;
    if (argc > 1) max_threads = std::atoi(argv[1]);

    // -----------------------------------------------------------------------
    printf("cpp-ipc benchmark\n");
    printf("Platform: ");
#if defined(__APPLE__)
    printf("macOS");
#elif defined(__linux__)
    printf("Linux");
#elif defined(_WIN32)
    printf("Windows");
#else
    printf("Unknown");
#endif
    printf(", %u hardware threads\n", std::thread::hardware_concurrency());

    // -----------------------------------------------------------------------
    print_header("ipc::route — 1 sender, N receivers (random 2-256 bytes x 100000)");
    printf("%10s  %12s  %12s\n", "Receivers", "RTT (ms)", "us/datum");
    printf("%10s  %12s  %12s\n", "----------", "----------", "----------");

    for (int n = 1; n <= max_threads; n *= 2) {
        auto s = bench_route(n, 100000, 2, 256);
        printf("%10d  %12.2f  %12.3f\n", n, s.total_ms, s.us_per_datum());
    }

    // -----------------------------------------------------------------------
    print_header("ipc::channel — 1-N (random 2-256 bytes x 100000)");
    printf("%10s  %12s  %12s\n", "Receivers", "RTT (ms)", "us/datum");
    printf("%10s  %12s  %12s\n", "----------", "----------", "----------");

    for (int n = 1; n <= max_threads; n *= 2) {
        auto s = bench_channel("1-N", n, 100000, 2, 256);
        printf("%10d  %12.2f  %12.3f\n", n, s.total_ms, s.us_per_datum());
    }

    // -----------------------------------------------------------------------
    print_header("ipc::channel — N-1 (random 2-256 bytes x 100000)");
    printf("%10s  %12s  %12s\n", "Senders", "RTT (ms)", "us/datum");
    printf("%10s  %12s  %12s\n", "----------", "----------", "----------");

    for (int n = 1; n <= max_threads; n *= 2) {
        auto s = bench_channel("N-1", n, 100000, 2, 256);
        printf("%10d  %12.2f  %12.3f\n", n, s.total_ms, s.us_per_datum());
    }

    // -----------------------------------------------------------------------
    print_header("ipc::channel — N-N (random 2-256 bytes x 100000)");
    printf("%10s  %12s  %12s\n", "Threads", "RTT (ms)", "us/datum");
    printf("%10s  %12s  %12s\n", "----------", "----------", "----------");

    for (int n = 1; n <= max_threads; n *= 2) {
        auto s = bench_channel("N-N", n, 100000, 2, 256);
        printf("%10d  %12.2f  %12.3f\n", n, s.total_ms, s.us_per_datum());
    }

    printf("\nDone.\n");
    return 0;
}

File: rust/libipc/src/bin/bench_ipc.rs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

use libipc::channel::{Channel, Mode, Route};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

struct Stats {
    total_ms: f64,
    count: usize,
}

impl Stats {
    fn us_per_datum(&self) -> f64 {
        (self.total_ms * 1000.0) / self.count as f64
    }
}

// ---------------------------------------------------------------------------
// ipc::route  —  1 sender, N receivers  (random msg_lo–msg_hi bytes × count)
// ---------------------------------------------------------------------------

fn bench_route(n_receivers: usize, count: usize, msg_lo: usize, msg_hi: usize) -> Stats {
    let name = "bench_route";

    let mut threads = Vec::new();

    // prepare random payloads
    let mut sizes = Vec::with_capacity(count);
    let mut rng_state: u64 = 42;
    for _ in 0..count {
        rng_state = rng_state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        let s = msg_lo + (rng_state >> 32) as usize % (msg_hi - msg_lo + 1);
        sizes.push(s);
    }
    let payload = vec![b'X'; msg_hi];

    // sender (created first so shm exists for receivers)
    Route::clear_storage(name);
    let mut sender = Route::connect(name, Mode::Sender).unwrap();

    let ready = Arc::new(AtomicBool::new(false));
    let done = Arc::new(AtomicBool::new(false));

    // receivers
    for _ in 0..n_receivers {
        let ready_clone = ready.clone();
        let done_clone = done.clone();
        threads.push(thread::spawn(move || {
            let mut r = Route::connect(name, Mode::Receiver).unwrap();
            while !ready_clone.load(Ordering::Acquire) {
                thread::yield_now();
            }
            while !done_clone.load(Ordering::Acquire) {
                let _buf = r.recv(Some(100));
            }
        }));
    }

    // let receivers connect
    thread::sleep(Duration::from_millis(100));
    ready.store(true, Ordering::Release);

    let t0 = Instant::now();

    for &size in &sizes {
        sender.send(&payload[..size], 0).unwrap();
    }

    let t1 = Instant::now();
    let total_ms = t1.duration_since(t0).as_secs_f64() * 1000.0;

    // signal done, disconnect sender to unblock receivers, then join
    done.store(true, Ordering::Release);
    sender.disconnect();
    for t in threads {
        t.join().unwrap();
    }

    Stats { total_ms, count }
}

// ---------------------------------------------------------------------------
// ipc::channel  —  pattern  (random msg_lo–msg_hi bytes × count)
//   pattern: "1-N"  = 1 sender,  N receivers
//            "N-1"  = N senders, 1 receiver
//            "N-N"  = N senders, N receivers
// ---------------------------------------------------------------------------

fn bench_channel(pattern: &str, n: usize, count: usize, msg_lo: usize, msg_hi: usize) -> Stats {
    let name = "bench_chan";

    let n_senders = if pattern == "N-1" || pattern == "N-N" { n } else { 1 };
    let n_receivers = if pattern == "1-N" || pattern == "N-N" { n } else { 1 };

    let per_sender = count / n_senders;

    // prepare random payloads
    let mut sizes = Vec::with_capacity(count);
    let mut rng_state: u64 = 42;
    for _ in 0..count {
        rng_state = rng_state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        let s = msg_lo + (rng_state >> 32) as usize % (msg_hi - msg_lo + 1);
        sizes.push(s);
    }
    let payload = vec![b'X'; msg_hi];

    // a "control" channel to keep shm alive; also used to disconnect receivers
    Channel::clear_storage(name);
    let mut ctrl = Channel::connect(name, Mode::Sender).unwrap();

    let ready = Arc::new(AtomicBool::new(false));
    let done = Arc::new(AtomicBool::new(false));

    let mut recv_threads = Vec::new();

    // receivers
    for _ in 0..n_receivers {
        let ready_clone = ready.clone();
        let done_clone = done.clone();
        recv_threads.push(thread::spawn(move || {
            let mut ch = Channel::connect(name, Mode::Receiver).unwrap();
            while !ready_clone.load(Ordering::Acquire) {
                thread::yield_now();
            }
            while !done_clone.load(Ordering::Acquire) {
                let _buf = ch.recv(Some(100));
            }
        }));
    }

    thread::sleep(Duration::from_millis(100));
    ready.store(true, Ordering::Release);

    let t0 = Instant::now();

    // senders
    let mut sender_threads = Vec::new();
    let sizes_arc = Arc::new(sizes);
    for s in 0..n_senders {
        let sizes_clone = sizes_arc.clone();
        let p = payload.clone();
        sender_threads.push(thread::spawn(move || {
            let mut ch = Channel::connect(name, Mode::Sender).unwrap();
            let base = s * per_sender;
            for i in 0..per_sender {
                ch.send(&p[..sizes_clone[base + i]], 0).unwrap();
            }
        }));
    }
    for t in sender_threads {
        t.join().unwrap();
    }

    let t1 = Instant::now();
    let total_ms = t1.duration_since(t0).as_secs_f64() * 1000.0;

    done.store(true, Ordering::Release);
    ctrl.disconnect();
    for t in recv_threads {
        t.join().unwrap();
    }

    Stats { total_ms, count }
}

fn print_header(title: &str) {
    println!("\n=== {} ===", title);
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let max_threads = if args.len() > 1 {
        args[1].parse().unwrap_or(8)
    } else {
        8
    };

    println!("cpp-ipc benchmark (Rust port)");
    let os = if cfg!(target_os = "macos") {
        "macOS"
    } else if cfg!(target_os = "linux") {
        "Linux"
    } else if cfg!(target_os = "windows") {
        "Windows"
    } else {
        "Unknown"
    };
    println!("Platform: {}, {} hardware threads\n", os, std::thread::available_parallelism().map(|n| n.get()).unwrap_or(1));

    // -----------------------------------------------------------------------
    print_header("ipc::route — 1 sender, N receivers (random 2-256 bytes x 100000)");
    println!("{:>10}  {:>12}  {:>12}", "Receivers", "RTT (ms)", "us/datum");
    println!("{:>10}  {:>12}  {:>12}", "----------", "----------", "----------");

    let mut n = 1;
    while n <= max_threads {
        let s = bench_route(n, 100000, 2, 256);
        println!("{:>10}  {:>12.2}  {:>12.3}", n, s.total_ms, s.us_per_datum());
        n *= 2;
    }

    // -----------------------------------------------------------------------
    print_header("ipc::channel — 1-N (random 2-256 bytes x 100000)");
    println!("{:>10}  {:>12}  {:>12}", "Receivers", "RTT (ms)", "us/datum");
    println!("{:>10}  {:>12}  {:>12}", "----------", "----------", "----------");

    let mut n = 1;
    while n <= max_threads {
        let s = bench_channel("1-N", n, 100000, 2, 256);
        println!("{:>10}  {:>12.2}  {:>12.3}", n, s.total_ms, s.us_per_datum());
        n *= 2;
    }

    // -----------------------------------------------------------------------
    print_header("ipc::channel — N-1 (random 2-256 bytes x 100000)");
    println!("{:>10}  {:>12}  {:>12}", "Senders", "RTT (ms)", "us/datum");
    println!("{:>10}  {:>12}  {:>12}", "----------", "----------", "----------");

    let mut n = 1;
    while n <= max_threads {
        let s = bench_channel("N-1", n, 100000, 2, 256);
        println!("{:>10}  {:>12.2}  {:>12.3}", n, s.total_ms, s.us_per_datum());
        n *= 2;
    }

    // -----------------------------------------------------------------------
    print_header("ipc::channel — N-N (random 2-256 bytes x 100000)");
    println!("{:>10}  {:>12}  {:>12}", "Threads", "RTT (ms)", "us/datum");
    println!("{:>10}  {:>12}  {:>12}", "----------", "----------", "----------");

    let mut n = 1;
    while n <= max_threads {
        let s = bench_channel("N-N", n, 100000, 2, 256);
        println!("{:>10}  {:>12.2}  {:>12.3}", n, s.total_ms, s.us_per_datum());
        n *= 2;
    }

    println!("\nDone.");
}
